### PR TITLE
Refactor BGC data loader

### PR DESCRIPTION
**Pull request by CunliangGeng**
This PR intends only to refactor BGC loaders. However, to make the genomics loading workflow works, I also updated some code for parsing bigscape and loading GCFs, this part is not fully refactored and will be done in next PR.

-  add abstract base class for BGC loader
-  update MIBiG BGC loader
-  add AntiSMASH BGC loader
-  update genomics load workflow `_load_genomics from loader.py` (not fully refactored yet, just to make sure it works, same below)
   - update workflow of loading BigSCAPE files and related BGCs `load_gcfs in genomics.py`
   -  update bigscape default parameter to make sure BGC has only one parent GCF
 - update bigscape installation with version 1.5.1, which supports mibig dataset v3.1
 - update logging

### CODE REVIEW

**Comment by CunliangGeng** [comment]
@hechth  ready for review :-)

**Comment by hechth** [comment]
@CunliangGeng 1891 files changed? I think something got checked in that shouldn't have?

File: .github/workflows/typing-check.yml [hunk 1]
```diff
@@ -35,6 +35,7 @@ jobs:
         run: |
           python3 -m pip install --upgrade pip setuptools
           pip install mypy
+          pip install -r requirements.dev.txt
```
Comment (hechth) [inline .github/workflows/typing-check.yml]: Why are the dev requirements necessary for the typing check?

**Comment by CunliangGeng** [comment]
> @CunliangGeng 1891 files changed? I think something got checked in that shouldn't have?

@hechth It's caused by AntiSMASH test data. I tried to mimic the scenario that users directly use AntiSMASH outputs, so I directly added two AntiSMASH output cases which turn out to have >1700 files. 

I have removed most AntiSMASH output files and kept enough to mimic the scenario.

File: pyproject.toml [hunk 2]
```diff
@@ -30,6 +30,7 @@ classifiers = [
 dependencies = [
     "beautifulsoup4",
     "biopython",
+    "deprecated",
```
Comment (hechth) [inline pyproject.toml]: Is there maybe a way to work around the dependencies being duplicated in the `requirements.txt` and the `pyproject.toml`?

File: src/nplinker/genomics/antismash/antismash_loader.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+logger = LogConfig.getLogger(__name__)
+
+
+class AntismashBGCLoader:
+
+    def __init__(self, data_dir: str) -> None:
+        """Build a loader for AntiSMASH BGC genbank (.gbk) files
+
+        Note:
+            AntiSMASH BGC direcotry must follow the strcuture below:
```
Comment (hechth) [inline src/nplinker/genomics/antismash/antismash_loader.py]:
```suggestion
            AntiSMASH BGC directory must follow the structure below:
```

File: src/nplinker/genomics/antismash/antismash_loader.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        bgc = BGC(-1,
+                  strain=strain,
+                  name=fname,
+                  product_prediction=product_prediction,
+                  description=description)
+
+    bgc.antismash_accession = antismash_accession
+    bgc.antismash_id = antismash_id
+
+    return bgc
```
Comment (hechth) [inline src/nplinker/genomics/antismash/antismash_loader.py:150]: Does this specific strain object already exist (with that name) or will it always only be instantiated there?

File: tests/test_utils.py [hunk 5]
```diff
@@ -128,15 +128,15 @@ def test_default(self):
...
 
     def test_prefix(self):
-        dirs = utils.list_dirs(root=self.root, prefix=True)
+        dirs = utils.list_dirs(root=self.root, keep_parent=False)
         assert isinstance(dirs, list)
         assert len(dirs) >= 1
-        assert "data" not in dirs
-        assert str(self.root / "data") in dirs
+        assert "data" in dirs
+        assert str(self.root / "data") not in dirs
```
Comment (hechth) [inline tests/test_utils.py]: A parameterized test that checks for `keep_parent=True` as well as `False` would be nice.

**Review by hechth — APPROVED** [review]
Seems good to me, just some minor remarks.

File: .github/workflows/typing-check.yml (hunk 1 above)
Comment (CunliangGeng) [reply .github/workflows/typing-check.yml]: not all dependencies but only `types-toml` -- the stubs for toml.
Changed it to only install `types-toml`.

File: src/nplinker/genomics/antismash/antismash_loader.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+from nplinker.logconfig import LogConfig
+from nplinker.utils import list_dirs
+from nplinker.utils import list_files
+from ..abc import BGCLoaderBase
+from nplinker.strains import Strain
+
+logger = LogConfig.getLogger(__name__)
+
+
+class AntismashBGCLoader:
```
Comment (hechth) [inline src/nplinker/genomics/antismash/antismash_loader.py:14]: Shouldn't this also implement the `BGCLoaderBase` class?

File: pyproject.toml (hunk 2 above)
Comment (CunliangGeng) [reply pyproject.toml]: yes, there is [a way](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata). Code updated.

File: src/nplinker/genomics/antismash/antismash_loader.py (hunk 4 above)
Comment (CunliangGeng) [reply src/nplinker/genomics/antismash/antismash_loader.py:150]: Good point!  I planned to solve this issue in next PR. The instantiation here should be removed, which will influence the logics of whole genomic loading process, so will solve it in next step.

File: tests/test_utils.py (hunk 5 above)
Comment (CunliangGeng) [reply tests/test_utils.py]: First time to use it :-)

File: src/nplinker/genomics/antismash/antismash_loader.py (hunk 6 above)
Comment (CunliangGeng) [reply src/nplinker/genomics/antismash/antismash_loader.py:14]: It did by registering as a virtual class of `BGCLoaderBase` (see the bottom of the file)

**Comment by CunliangGeng** [comment]
code updated as what you suggested, take a look please

**Pull request merged**